src/App/VPSScan/Scan.hs

```diff
@@ -36,37 +37,49 @@ scanMain opts@ScanCmdOpts{..} = do
 ----- main logic
 
 vpsScan ::
-  ( Has Diagnostics sig m
-  , Has Trace sig m
+  (Has Trace sig m
   , MonadIO m
+  , Has Diagnostics sig m
   ) => Path Abs Dir -> ScanCmdOpts -> m ()
-vpsScan basedir ScanCmdOpts{..} = do
-  let vpsOpts@VPSOpts{..} = scanVpsOpts
-  response <- context "creating scan ID" $ createScotlandYardScan vpsOpts
+vpsScan basedir opts@ScanCmdOpts{..} = do
+  response <- context "creating scan ID" $ createScotlandYardScan scanVpsOpts
   let scanId = responseScanId response
-
   trace $ "Running scan on directory " ++ show basedir
   trace $ "Scan ID from Scotland yard is " ++ show scanId
   trace "[All] Running IPR and Sherlock scans in parallel"
   trace "[Sherlock] Starting Sherlock scan"
+  _ <- runScans basedir scanId opts
+  trace "[All] All scans complete"
+
+runScans :: (Has Diagnostics sig m, Has Trace sig m, MonadIO m) => Path Abs Dir -> Text -> ScanCmdOpts -> m ()
+runScans basedir scanId ScanCmdOpts{..} = do
+  let vpsOpts@VPSOpts{..} = scanVpsOpts
   case vpsIpr of
     Just _ -> trace "[IPR] Starting IPR scan"
     Nothing -> trace "[IPR] IPR scan disabled"
 
-  let runIt = runDiagnostics . runExecIO . runTrace
+  let runIt = runExecIO . runTrace . runDiagnostics
   (iprResult, sherlockResult) <- liftIO $ concurrently
-                (runIt $ runIPRScan basedir scanId vpsOpts)
-                (runIt $ runSherlockScan basedir scanId vpsOpts)
+         (runIt $ runIPRScan basedir scanId vpsOpts)
+         (runIt $ runSherlockScan basedir scanId vpsOpts)
+  trace "[All] Scans complete"
   case (iprResult, sherlockResult) of
-    (Right _, Right _) -> trace "[All] Scans complete"
     (Left iprFailure, _) -> do
-      trace "[IPR] Failed to scan"
-      trace (show $ renderFailureBundle iprFailure)
+      liftIO $ print $ renderFailureBundle iprFailure
       liftIO exitFailure
     (_, Left sherlockFailure) -> do
-      trace "[Sherlock] Failed to scan"
-      trace (show $ renderFailureBundle sherlockFailure)
+      liftIO $ print $ renderFailureBundle sherlockFailure
       liftIO exitFailure
+    (_, _) ->
```

Comment:
[nit] this is the same as just `_`

Reply:
also - because `exitFailure` will just exit immediately, we don't have to nest these case expressions - we can use a new top-level case